### ISSUE TITLE
Second scrollbar in application layout if content is longer than the viewport

### ISSUE THREAD

**Issue by steverydz** [canonical-web-and-design/vanilla-framework#3440]
**Describe the bug**

When using the application layout, if the page content is longer than the viewport, a second scrollbar appears for the main panel of the layout.

![application_layout_double_scrollbar](https://user-images.githubusercontent.com/501889/100887310-b9228a80-34ac-11eb-9776-04f078d694fc.png)

**Comment by bartaz** [comment]
Application layout currently doesn't support navigation bar on top of whole screen. Not sure what markup you used to get it there, but it's likely cause of additional scrollbar.

@steverydz  Can you provide the example with HTML to debug? But I can't promise we can support such exception to the designed layout at this point.

**Comment by steverydz** [comment]
> Application layout doesn't support navigation bar on top of whole screen. Not sure what markup you used to get it there, but it's likely cause of additional scrollbar.

It only looks as if it's the whole width because of the smaller viewport. Without dev tools open it looks like this: https://user-images.githubusercontent.com/501889/100896976-1e7b7900-34b7-11eb-9a17-934a598bdcf8.png

Here's the markup for the page: https://pastebin.canonical.com/p/2JGtnPCgFP/

**Comment by bartaz** [comment]
@steverydz Thanks.

In the application layout `l-application` is supposed to be only and direct child of body element as it's meant to be full screen container for application panels. In your case additional header with navigation seems to push it down.
`l-application` has height set to `100vh`, so it will be the size of the viewport even if something is above it, causing the scrollbar to appear.

It doesn't seem to be a bug. Application layout was not designed to have a header like this. If we are to support it, we can have it discussed as pattern amendment.

At this point I don't really have any good suggestion how to workaround this. `l-application` uses CSS grid layout, so the only way to make this work is to modify the grid definition in `l-application` to have additional area for header (on top of everything else) and put this kind of navigation there. But it may be quite complicated to hack it into as custom local styles, as the grid changes a bit on different screen sizes so all of that would need to be taken into account.

**Comment by lyubomir-popov** [comment]
Why use the application layout here to begin with? This looks like a normal site.

**Comment by steverydz** [comment]
> Why use the application layout here to begin with? This looks like a normal site.

That's the layout used in the designs: https://app.zeplin.io/project/5f3170425822f41fc299a330/screen/5fb6913708e04b418dfbb8a3

**Comment by anthonydillon** [comment]
Could you ask if the to navigation could be added to the side navigation?

**Comment by lyubomir-popov** [comment]
> That's the layout used in the designs:

Ah got it. `p-navigation__row--full-width` will stretch the navigation. Isn't the rest possible without the application layout? You may need to define a width for the side nav, other than that I don't see anything else that would need to be custom.

**Comment by lyubomir-popov** [comment]
@bartaz @sowasred2012 Seems there's a need for a second application layout has the following features:

- header spanning full width, height defined by content, sticky
- left sidebar
- main area
- optional right panels, status bar

In terms of naming, maybe we can have `--horisontal` and the current one being default or `--vertical`?

Could looke like so:
![Uploading image.png…]()

**Comment by bartaz** [comment]
@lyubomir-popov Shouldn't we try to make apps consistent? Is their use case that different that needs separate layout? Will we use this layout elsewhere?

We could try to make this part of application layout (an optional top header area?), but obviously the more optional elements the harder it will be to make it work in all supported cases.

**Comment by lyubomir-popov** [comment]
There are valid needs for both. Here's an example from microsoft:

![image](https://user-images.githubusercontent.com/2741678/101024715-6ad4c080-356c-11eb-81cc-545ed894d4fe.png)
![image](https://user-images.githubusercontent.com/2741678/101024769-7b853680-356c-11eb-9f8a-bfc5c6ef5613.png)


The choice depends on multiple factors. like number of items in the menu, whether the app needs the full width of the window, having to stay consistent with a brochure site, etc. We agreed in design review to define criteris for choosing one over the other and explain it in the documentation. (some notes [here](https://docs.google.com/document/d/1YuEQLWLKaf-Rxbn0g1St-FdsrCnTAKihVoBWZX_60M4/edit#bookmark=id.5s2m7p8mscad)

In terms of whether this should be a separate layou vs  an optional area, I'd vote for separate layout. Much easier to point someone at the correct example they can copy>paste, rather than requiring them learning how to configure one master example to suit their needs.

As the layouts evolve, they will have more and more things thst are tailored to each, trying to keep those as configuration options on the same layout would be harder I think.

**Comment by bartaz** [comment]
If it's separate layout, let's call it different, not `application--something`. This suggests it's just a variant of application layout (like a `--deep` variant of strip). If we consider it separate, it shouldn't be called application layout.

**Comment by lyubomir-popov** [comment]
@bartaz I guess "application" is the fuzzy part - as we have a lot of hybrids between site and application. 

In the design meeting it became clear that we might understand the term application layout slightly diffferently from design and dev perspective. It could mean, among other things:

- overall position of landmark areas like main nav
- scroll behaviour (stick, scroll within container, etc)

So a horisontal top nav layout and a vertical left nav layout will share some aspects but differ in others. Exactly what terminology we use is probably something to be decided in a joint manner between design/dev to reduce ambiguity.

CC @clagom  @spencerbygraves

**Comment by spencerbygraves** [comment]
A few questions...

Currently in the Snap admin dashboard, both top and side nav have been deemed necessary. What happens with separate layouts in that instance?

Could separate layouts become too fixed? When we have another new requirement such as "this layout needs the top nav plus XYZ", will we need layout 3?

Do we perhaps need a set of parts to use and UX guidance for creating apps?

**Comment by lyubomir-popov** [comment]
> Currently in the Snap admin dashboard, both top and side nav have been deemed necessary. What happens with separate layouts in that instance?

I would propose the following distinction:

#### What occupies the top left corner? 
- If the horisontal nav, then that is the primary nav, the rest is simply page content. 
- If it is the left sidenav, then this is the primary nav, and the rest is page content that corresponds to the currently selected menu item in the side nav

**Comment by bartaz** [comment]
Does the snapcraft's side navigation behave like in application layout? Is it meant to collapse on smaller screens, expand on larger, etc, as in application layout and JAAS dashboard?

Obviously the more similarities are between the two, the better would be for them to be one layout with optional elements.

The initial idea behind building layouts (and application layout as one of them) was to provide consistent look and feel across our products. Currently it seems instead of this being consistent layout for everyone, we seem to have a need for flexibility and exceptions: JAAS dashboard needs a top toolbar, above other panels, but next to side navigation. Snapcraft needs top level global navigation on top of everything... 

To build it as part of a framework we need some defined structure there. If we don't want to have one strict consistent layout but building blocks that can be used to compose ad-hoc solutions, we shouldn't build layouts, but those building blocks.

**Comment by clagom** [comment]
JAAS always had that top bar that was on top of panels, and arguably it would have been a requirement to feed back the pattern. I think we *all* just missed this and overlooked that part during the Application layout definition. The goal of a pattern is consistency, but that cannot be "apply exactly the same solutions" everywhere, that's why we got modularity in the table, modularity in the search and filter, and so on. I think the Application layout also needs a bit of flexibility if it wants to become the standard pattern across products/projects. We're not talking about different pattern, but top-nav yes/no, tool-bar yes/no, kinda thing. We need design/UX work on the definition of this modularity. We just released the Application layout, some adjustments and iterations are to be expected.
To get back to the original topic, @lilyvidenova has a better idea if the top-navigation here is necessary. Might be that when you go from snapcraft.io to snapcraft.io/brand-store (example url) then you are in the app view, similar to what currently happens between jaas.ai and jaas.ai/models.
